[quant][graphmode] Add RemoveReduantDequantize pass

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #38552 [quant] Make MinMaxObserver and MovingAverageMinMaxObserver work with List[Tensor]
* #38436 [quant] Histogram Observer support List[Tensor]
* #38570 [quant][graphmode] Different rule for handling `aten::cat`
* #38435 [quant] Remove get_qparams in Observers
* **#38434 [quant][graphmode] Add RemoveReduantDequantize pass**

Summary:
We insert dequantize for each use in order to produce quantization patterns that will
later be fused, after that we should also remove extra dequantize node produced by this operation.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21597834](https://our.internmc.facebook.com/intern/diff/D21597834)

jerryzh168 committed:
[quant][graphmode] Add RemoveReduantDequantize pass

Summary:
We insert dequantize for each use in order to produce quantization patterns that will
later be fused, after that we should also remove extra dequantize node produced by this operation.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Add RemoveReduantDequantize pass"


Summary:
We insert dequantize for each use in order to produce quantization patterns that will
later be fused, after that we should also remove extra dequantize node produced by this operation.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Add RemoveReduantDequantize pass"


Summary:
We insert dequantize for each use in order to produce quantization patterns that will
later be fused, after that we should also remove extra dequantize node produced by this operation.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar commented on test/quantization/test_quantize_script.py:
```diff
@@ -1015,6 +1016,22 @@ def forward(self, x):
+            def forward(self, x):
+                y = x.size(0)
+                x = self.conv(x)
+                return (x.size(0) + y) * x
```
Why does this case have extra dequant? Is it due to x.size()?

jerryzh168 replied:
Yes

jerryzh168 commented:
I thought about this again, and feel that this might not be very safe, since user can explicitly call dequantize as well, although that's rare, to be safe, I'll create a whitelist for ops that the dequantize can be removed for their input.

jerryzh168 committed:
Update on "[quant][graphmode] Add RemoveReduantDequantize pass"


Summary:
We insert dequantize for each use in order to produce quantization patterns that will
later be fused, after that we should also remove extra dequantize node produced by this operation.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Add RemoveReduantDequantize pass"


Summary:
We insert dequantize for each use in order to produce quantization patterns that will
later be fused, after that we should also remove extra dequantize node produced by this operation.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21597834](https://our.internmc.facebook.com/intern/diff/D21597834)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Add RemoveReduantDequantize pass"


Summary:
We insert dequantize for each use in order to produce quantization patterns that will
later be fused, after that we should also remove extra dequantize node produced by this operation.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21597834](https://our.internmc.facebook.com/intern/diff/D21597834)

[ghstack-poisoned]
